crates/lib/src/qvm/api.rs

```diff
@@ -0,0 +1,141 @@
+//! This module provides functions to for makings calls to the QVM.
+use std::borrow::Cow;
+use std::str::FromStr;
+
+use qcs_api_client_common::ClientConfiguration;
+use quil_rs::{
+    instruction::{ArithmeticOperand, Instruction, MemoryReference, Move},
+    Program,
+};
+
+use super::{Error, QvmResultData};
+use crate::{
+    executable::Parameters,
+    qvm::{Request, Response},
+};
+
+/// Execute a program on the QVM.
+pub async fn run(
+    quil: &str,
+    shots: u16,
+    readouts: &[Cow<'_, str>],
+    params: &Parameters,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    #[cfg(feature = "tracing")]
+    tracing::debug!(
+        %shots,
+        ?readouts,
+        ?params,
+        "preparing program to run on QVM"
+    );
+
+    let program = Program::from_str(quil).map_err(Error::Parsing)?;
+
+    run_program(&program, shots, readouts, params, config).await
+}
+
+pub(crate) async fn run_program(
+    program: &Program,
+    shots: u16,
+    readouts: &[Cow<'_, str>],
+    params: &Parameters,
+    config: &ClientConfiguration,
+) -> Result<QvmResultData, Error> {
+    if shots == 0 {
+        return Err(Error::ShotsMustBePositive);
+    }
+
+    // Create a clone of the program so MOVE statements can be prepended to it
+    let mut program = program.clone();
```

Comment:
This is a change in logic from before. This function took a `&mut Program`, prepended the move instructions then removed them again before returning.

My thinking is, for QVM runs at least, cloning the program makes the function signature more clear (in that the function wont change the program). It's also safer, as we can't accidentally modify the users Program without undoing the changes. Benchmark needed, but I would suspect for up to a certain size of program, working off the clone is faster, as you avoid resizing the vector for each instruction you need to peel off at the end (though that could be optimized with changes to quil-rs).

Thoughts?

Reply:
Good move IMO.

I would be surprised if the change in execution time would be noticeable in context of the rest of QVM duration